core/web/user_controller.go

```diff
@@ -28,6 +29,181 @@ type UpdatePasswordRequest struct {
 	NewPassword string `json:"newPassword"`
 }
 
+// Index lists all API users
+func (c *UserController) Index(ctx *gin.Context) {
+	users, err := c.App.SessionORM().ListUsers()
+	if err != nil {
+		c.App.GetLogger().Errorf("Unable to list users", "err", err)
+		jsonAPIError(ctx, http.StatusInternalServerError, err)
+		return
+	}
+	jsonAPIResponse(ctx, presenters.NewUserResources(users), "users")
+}
+
+// Create creates a new API user with provided context arugments.
+func (c *UserController) Create(ctx *gin.Context) {
+	type newUserRequest struct {
+		Email    string `json:"email"`
+		Password string `json:"password"`
+		Role     string `json:"role"`
+	}
+
+	var request newUserRequest
+	if err := ctx.ShouldBindJSON(&request); err != nil {
+		jsonAPIError(ctx, http.StatusUnprocessableEntity, err)
+		return
+	}
+
+	user, err := clsession.NewUser(request.Email, request.Password, request.Role)
+	if err != nil {
+		jsonAPIError(ctx, http.StatusBadRequest, fmt.Errorf("error creating API user: %s", err))
+		return
+	}
+	if err = c.App.SessionORM().CreateUser(&user); err != nil {
+		// If this is a duplicate key error (code 23505), return a nicer error message
+		if err, ok := err.(*pgconn.PgError); ok {
+			fmt.Println(err.Code)
+			if err.Code == "23505" {
+				jsonAPIError(ctx, http.StatusBadRequest, fmt.Errorf("user with email %s already exists", request.Email))
+				return
+			}
+		}
+		c.App.GetLogger().Errorf("Error creating new API user", "err", err)
+		jsonAPIError(ctx, http.StatusInternalServerError, fmt.Errorf("error creating API user"))
+		return
+	}
+
+	jsonAPIResponse(ctx, presenters.NewUserResource(user), "user")
+}
+
+// Update changes sets email, password, or role fields of a specified API user.
+func (c *UserController) Update(ctx *gin.Context) {
+	type updateUserRequest struct {
+		Email       string `json:"email"`
+		NewEmail    string `json:"newEmail"`
+		NewPassword string `json:"newPassword"`
+		NewRole     string `json:"newRole"`
+	}
+
+	var request updateUserRequest
+	if err := ctx.ShouldBindJSON(&request); err != nil {
+		jsonAPIError(ctx, http.StatusUnprocessableEntity, err)
+		return
+	}
+
+	// First, attempt to load specified user by email
+	// This user object will have its fields patched for a final update call
+	user, err := c.App.SessionORM().FindUser(request.Email)
+	if err != nil {
+		jsonAPIError(ctx, http.StatusBadRequest, fmt.Errorf("specified user not found: %s", request.Email))
+		return
+	}
+
+	// Don't allow current admin user to edit self
+	sessionUser, ok := webauth.GetAuthenticatedUser(ctx)
```

Comment:
What do you think about wrapping everything here in a transaction right within the controller function? Otherwise I can move the DB actions into the ORM method (fetch user, delete session, update user)

Reply:
Yes, this logic is actually pretty janky (it's very old). It smells off to me, we lookup the user record multiple times for example, and the session creation looks like it can race. We can sync offline about how to deal with that

Reply:
This has been addressed 🙂